File: src/pvz_console/core/diff.py
from __future__ import annotations

from typing import Iterable, List, Protocol, TypeVar


class _HasId(Protocol):
    id: object


T = TypeVar("T", bound=_HasId)


def missing_by_id(source: Iterable[T], target: Iterable[T]) -> List[T]:
    """Entries from ``source`` whose ``id`` does not appear in ``target``."""
    target_ids = {item.id for item in target}
    return [item for item in source if item.id not in target_ids]

File: src/core/diff.ts
import type { Id } from "./models";

export interface HasId {
  id: Id;
}

/** Entries from `source` whose `id` does not appear in `target`. */
export function missingById<T extends HasId>(source: Iterable<T>, target: Iterable<T>): T[] {
  const targetIds = new Set<Id>();
  for (const item of target) targetIds.add(item.id);
  return [...source].filter((item) => !targetIds.has(item.id));
}
